libs/injection/bbhnet/injection/injection.py

```diff
@@ -0,0 +1,261 @@
+        waveform_arguments={
+            "waveform_approximant": "IMRPhenomPv2",
+            "reference_frequency": 50,
+            "minimum_frequency": 20,
+        },
+    )
+
+    # sample GW parameters from prior distribution
+    priors = bilby.gw.prior.BBHPriorDict(prior_file)
+    sample_params = priors.sample(n_samples)
+
+    # Come back to this and see if can add frame_start
+    sample_params["geocent_time"] = signal_times
+
+    signal_strains = []
+    signals_list = []
+    snr_list = []
+    for frame, channel, ifo in zip(frame_files, channels, ifos):
+        # Get the strain from each frame file
+        strain = TimeSeries.read(frame, channel)
+
+        # calculate the PSD
+        strain_psd = strain.psd(fftlength)
+
+        # generate GW waveforms
+        signals, snr = generate_gw(
+            sample_params,
+            ifo,
+            waveform_generator=waveform_generator,
+            get_snr=True,
+            noise_psd=strain_psd,
+        )
+
+        signal_strains.append(strain)
+        signals_list.append(signals)
+        snr_list.append(snr)
+
+    old_snr = np.sqrt(np.sum(np.square(snr_list)))
+    new_snr = np.random.uniform(snr_range[0], snr_range[1], len(snr_list[0]))
+
+    signals_list = [
+        signals * (new_snr / old_snr)[:, None] for signals in signals_list
+    ]
+    sample_params["luminosity_distance"] = (
+        sample_params["luminosity_distance"] * old_snr / new_snr
+    )
+    snr_list = [snr * new_snr / old_snr for snr in snr_list]
+
+    for strain, signals, frame in zip(
+        signal_strains, signals_list, frame_files
+    ):
+        for i in range(n_samples):
+            idx1 = int(
+                (signal_times[i] - waveform_duration / 2.0) * sample_rate
+            )
+            idx2 = idx1 + waveform_duration * sample_rate
+            strain[idx1:idx2] += signals[i]
+
+        strain.write(os.path.join(outdir, frame))
+
+    # Write params and similar to output file
+    param_file = os.path.join(
+        outdir, f"param_file_{frame_start}-{frame_stop}.h5"
+    )
+    with h5py.File(param_file, "w") as f:
+        # write signals attributes, snr, and signal parameters
+        params_gr = f.create_group("signal_params")
+        for k, v in sample_params.items():
+            params_gr.create_dataset(k, data=v)
+
+        # Save signal times as actual GPS times
+        f.create_dataset("GPS-start", data=signal_times + frame_start)
+
+        for i, ifo in enumerate(ifos):
+            ifo_gr = f.create_group(ifo)
+            ifo_gr.create_dataset("signal", data=signals_list[i])
+            ifo_gr.create_dataset("snr", data=snr_list[i])
+
+        # write frame attributes
+        f.attrs.update(
+            {
+                "size": n_samples,
+                "frame_start": frame_start,
+                "frame_stop": frame_stop,
+                "sample_rate": sample_rate,
+                "psd_fftlength": fftlength,
+            }
+        )
```

Comment:
Part of me thinks this metadata might make more sense to store with each individual injection, and save them as HDF5 files rather than GWFs.
- If we do ~1e4-1e5 injections, this file could become huge
- Generalizes better to downstream applications that might just need a random filename without having to worry about the intended structure of the directory that this file lives in
- It looks like this uses the frame start of the first file that got passed, but it's not obvious how other files could map that to their own frame start time, especially since we don't actually enforce that the first file in `frame_files` is chronologically the first.

Then again, I don't fully know how this info gets used so maybe there's good reasons to prefer this. I'm also sympathetic to the idea of leaving frame files "as-is", basically identical except that there's some injections in them.

The other thing that strikes me is that if the SNR re-weighting step above is absolutely necessary to do after all the individual SNRs have been calculated, we would have to loop through all the already written files and add the reweighted SNR value to the corresponding `ifo` group.

Thoughts here?

Reply:
Opted against modification after conversation